docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+By keeping this information with the MachineSet and MachineDeployment resource objects, we reduce the
+need for the cluster autoscaler to have a deep knowledge of Cluster API internals.
+
+### User Stories
+
+- Detail the things that people will be able to do if this proposal is implemented.
+- Include as much detail as possible so that people can understand the "how" of the system.
+- The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
+cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
+a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.
+
+#### Story 2
+
+As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
+without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
+not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
+creation of nodes that will not consume resources until they are required by applications on my cluster.
+
+### Implementation Details/Notes/Constraints
+
+A new common API type will be created named `ResourceHints`, its definition will look like this:
+
+```
+type ResourceHints struct {
+    vCPU   *string `json:"vCPU,omitempty"`
+    Memory *string `json:"memory,omitempty"`
+    GPU    *string `json:"gpu,omitempty"`
+}
+```
+
+This new API type can be used with a machine template to inform about the resources used
+by individual machines in the group. For example, here is a `DockerMachineTemplate` with
+the integrated `ResourceHints`
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  resourceHints:
+    memory: 500mb
+    vCPU: "1"
+    GPU: "1"
+  template:
+    spec: {}
+```
+
+During reconciliation of MachineSets and MachineDeployments, if the `resourceHints` field is found
+on a related `infrastructureRef`, then the `resourceHints` will be added to the `status` field
+for that resource. For example, a MachineDeployment status might look like this:
+
+```
+status:
+  observedGeneration: 1
+  phase: Running
+  selector: cluster.x-k8s.io/cluster-name=workload,cluster.x-k8s.io/deployment-name=workload-md-0
+  resourceHints:
+    memory: 500mb
+    vCPU: "1"
+    GPU: "1"
+```
+
+If a user wishes to override the resources hints as provided from the machine template, they may
+do so by adding the following annotations to their MachineSets or MachineDeployments:
+
+```
+kind: MachineSet
+metadata:
+  annotations:
+      machine.x-k8s.io/GPU: "0"
```

Comment:
Also in this case, wondering if the annotation should refer to the autoscaler explicitly

Reply:
i think that makes sense, i'll think about some name ideas.